File: BillingSystem.Sales/Domain/ValueObjects/Percentage.cs
namespace BillingSystem.Sales.Domain.ValueObjects;

public class Percentage
{
    public decimal Value { get; }

    public Percentage(decimal value)
    {
        if (value < 0 || value > 100)
            throw new ArgumentException("Percentage must be between 0 and 100.");
        Value = value;
    }

    public static Percentage operator +(Percentage p1, Percentage p2)
    {
        return new Percentage(p1.Value + p2.Value);
    }

    public static Percentage operator -(Percentage p1, Percentage p2)
    {
        return new Percentage(p1.Value - p2.Value);
    }

    public override string ToString() => Value.ToString("0.0000");
}

File: BillingSystem.Sales/Client/src/app/models/Percentage.ts
export class Percentage {
    public value: number;

    constructor(value: number) {
        if (value < 0 || value > 100) throw new Error("Percentage must be between 0 and 100.");
        this.value = value;
    }

    public static add(p1: Percentage, p2: Percentage): Percentage {
        return new Percentage(p1.value + p2.value);
    }

    public static subtract(p1: Percentage, p2: Percentage): Percentage {
        return new Percentage(p1.value - p2.value);
    }
}
